### ISSUE TITLE
Element-R: Login-with-SSSS

### ISSUE THREAD

**Issue by florianduros** [vector-im/element-web#25648]
https://docs.google.com/document/d/18LuSmlaoeZpsOsiPKIdAxKYsbr9V1etMAYm2vxMId7A/edit#heading=h.a2i6v049d8mf
```[tasklist]
### Tasks
- [ ] https://github.com/matrix-org/matrix-react-sdk/pull/11217
- [ ] https://github.com/matrix-org/matrix-js-sdk/pull/3552
- [ ] https://github.com/matrix-org/matrix-rust-sdk/pull/2229
- [x] https://github.com/matrix-org/matrix-js-sdk/pull/3546 CANCELLED
- [ ] https://github.com/matrix-org/matrix-js-sdk/pull/3608
- [ ] https://github.com/matrix-org/matrix-js-sdk/pull/3619
- [ ] https://github.com/matrix-org/matrix-react-sdk/pull/11323
```

**Comment by richvdh** [comment]
`checkOwnCrossSigningTrust` is ... confusing.

The legacy implementation is [here](https://github.com/matrix-org/matrix-js-sdk/blob/v26.2.0/src/crypto/index.ts#L1575). It is called (directly or via the shim in `MatrixClient` from the following places:
 * In legacy `bootstrapCrossSigning` (with `allowPrivateKeyRequests: true`), when the private cross-signing keys are already in secret storage.
 * Whenever we re-download the user's e2e keys, when the cross-signing keys have *not* changed (with `allowPrivateKeyRequests: false`). I strongly suspect this is redundant and it got cargo-culted in https://github.com/matrix-org/matrix-js-sdk/commit/a34758f938930977156ed94b94412ada99746666#diff-de8313e2e1eba20bb7f59d6a94b95837a75fd4eb39d0ba0e5bd34e04d80de75cR530-R549.
 * In `tryToUnlockSecretStorageWithDehydrationKey` (with `allowPrivateKeyRequests: false`). We'll ignore this for now since it's related to dehydration.
 * In [`SetupEncryptionStore.usePassphrase`](https://github.com/matrix-org/matrix-react-sdk/blob/v3.75.0/src/stores/SetupEncryptionStore.ts#L147) (with `allowPrivateKeyRequests: false`). `SetupEncryptionStore.usePassphrase` is used when the user clicks "Verify with Security Key or Phrase" in the encryption setup dialog.


So, what does it actually do?

1. Checks we have downloaded the user's public e2e keys at least once, and waits for any in-flight key downloads to complete.
2. Attempts to read the private cross-signing keys from indexeddb (`crossSigningInfo.getCrossSigningKeysFromCache()`)
3. If (1) `allowPrivateKeyRequests: true` (ie, this is from `bootstrapCrossSigning`) and (2) either the in-memory public master cross-signing key does not match that in the `/keys/query` result, or there is no private cross-signing key in indexedDB:
   
   Call `CrossSigning.getCrossSigningKey`. The documentation for this simply says "Calls the app callback to ask for a private key" but actually we throw away the result so we appear to be calling for its undocumented side-effects, which is to **store the returned key in indexedDB** (as `account.ssss_cache:master`)

   However, `matrix-react-sdk` does not provide a `getCrossSigningKey` callback, so we get a [default](https://github.com/matrix-org/matrix-js-sdk/blob/v26.2.0/src/crypto/index.ts#L512-L516) which reads it from SSSS.

   TLDR: this reads the private cross-signing key from SSSS and stores it in indexeddb.

4. `storeTrustedSelfKeys`. Checks that the public user-signing and self-signing keys are signed by the master key, and then updates the in-memory and in-indexedDB (`account.crossSigningKeys`) cache of our *public* cross-signing keys. (Note that any device signed by these keys is then considered trusted, hence the name).

5. Do a similar dance to 3, for the user-signing and self-signing keys.
6. If 5 produced a new private self-signing key, sign the current device with it.
7. If there is a new master cross-signing key, sign it with the current device (whether or not we have the private key)
8. Upload the signatures produced by 6 and 7.
9. Emit some events
10. If there is a new master cross-signing key, call `afterCrossSigningLocalKeyChange`, which:
    * Signs the device and uploads the signature *again*!
    * If there is a `shouldUpgradeDeviceVerifications` callback, do some stuff. But there isn't, for matrix-react-sdk.


In summary then, this method is responsible for:
 * for `bootStrapSecretStorage` only, importing private cross-signing keys from secret storage.
 * updating our idea of the current (verified) public cross-signing keys
 * signing the master-key with the device and the device with the self-signing key.

**Comment by richvdh** [comment]
I'm struggling to figure out what the equivalents for those operations are in the Rust SDK. Currently:
 * "importing private cross-signing keys from secret storage" is basically [`OlmMachine::import_cross_signing_keys`](https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk_crypto/struct.OlmMachine.html#method.import_cross_signing_keys). But that doesn't do any signing, nor does it mark the imported identity as verified, as far as I can tell.
 * You can mark the identity as verified manually via [`OwnUserIdentity::mark_as_verified`](https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk_crypto/struct.OwnUserIdentity.html#method.mark_as_verified)
 * Alternatively you can use `OwnUserIdentifty::verify()` to do the same thing and add a signature of the master key. But that doesn't sign the devies with the self-signing key>
 * There is some logic for calling `OwnUserIdentity::mark_as_verified` when a new public key is received that matches the current private key - but it doesn't issue any signatures, and it doesn't fire when a new private key is imported that matches the current public key.

**Comment by richvdh** [comment]
Of those four call sites (`bootstrapCrossSigning`, `_onDeviceListUserCrossSigningUpdated`, `tryToUnlockSecretStorageWithDehydrationKey`, `SetupEncryptionStore.usePassphrase`), as I already said `_onDeviceListUserCrossSigningUpdated` and `_onDeviceListUserCrossSigningUpdated` can be ignored for now.

Of the remaining two: the call in `SetupEncryptionStore.usePassphrase` is used in a callback to `SecurityManager.accessSecretStorage`, which uses it [here](https://github.com/matrix-org/matrix-react-sdk/blob/v3.75.0/src/SecurityManager.ts#L392). Note that the callback follows -  - either directly, or via `CreateSecretStorageDialog` - a call to `bootstrapCrossSigning`.

In other words - I'm 90% sure that the call in `SetupEncryptionStore.usePassphrase` is redundant.

I think the confusion arose because `bootstrapCrossSigning` is called before `bootstrapSecretStorage`, so you might question how the cross-signing key can be imported from secret storage before secret storage is set up. However, in practice, even if `bootstrapSecretStorage` has not been called, `CrossSigning.getCrossSigningKey` calls the application-level `getFromSecretStorage` callback, which opens the `AccessSecretStorageDialog` to prompt for the SSSS key and thus enables access.

Assuming the call in `usePhrase` *is* redundant, This is good, because it makes the interface much easier to deal with. In order to make that certainty 100%, let's stub out `checkOwnCrossSigningTrust` in `RustCrypto` for now, and instead update `RustCrypto.CrossSigningIdentity.bootstrapCrossSigning` to call an internal method to import the private keys from SSSS.

**Comment by florianduros** [comment]
Despite the https://github.com/matrix-org/matrix-js-sdk/pull/3608 PR, the tests are sill in error when running against the old crypto. However they are failing later in the verification stage.

After the secret storage access, the dialog is stuck in the verifying stage. This is how it works in the old crypto:
1. https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/structures/auth/CompleteSecurity.tsx#L46 `phase` is set as `Phase.Done` => The verification is considered as terminated
2. The store event is raised https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/stores/SetupEncryptionStore.ts#L177 here
3. Which is triggered by the `CryptoEvent.UserTrustStatusChanged` event https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/stores/SetupEncryptionStore.ts#L68
4. The `CryptoEvent.UserTrustStatusChanged` is fired by `Crypto.checkOwnCrossSigningTrust` in the old crypto https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/crypto/index.ts#L1743
5. The `Crypto.checkOwnCrossSigningTrust` function is called in `Crypto.bootstrapCrossSigning` https://github.com/matrix-org/matrix-js-sdk/blob/develop/src/crypto/index.ts#L862
6. Additionally https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/stores/SetupEncryptionStore.ts#L174 `CryptoAPI.getCrossSigningKeyId` is not implemented in the rust-crypto

In conclusion, the rust-crypto doesn't fire the `CryptoEvent.UserTrustStatusChanged` event and doesn't implement the `CryptoAPI.getCrossSigningKeyId` function. However I think we want to get rid of the `checkOwnCrossSigningTrust` in the new crypto and its event.
